GH-3047: Add method and args headers into gateway

artembilan opened the pull request:
Fixes https://github.com/spring-projects/spring-integration/issues/3047

In some use-cases we can't use `headerExpressions` for messaging gateway
interface, e.g. when we use third-party class and don't use XML, but
Java DSL.

* Always add `gatewayMethod` and `gatewayArgs` headers to messages
produced by the gateway calls.
This information may be useful for downstream routing or analyzes
* Demonstrate a routing approach in the Docs with Java DSL.
Expose convenient methods to access those headers in the
`StaticMessageHeaderAccessor`
* Improve `GatewayProxyFactoryBean` to determine the return type of the
method call from the interface generic types, when the `serviceInterface`
is a `java.util.function.Function`

<!--
Thanks for contributing to Spring Integration. Please provide a brief description of your pull-request and reference any related issue numbers (prefix references with #).

See the [Contributor Guidelines for more information](https://github.com/spring-projects/spring-integration/blob/master/CONTRIBUTING.adoc).
-->

artembilan committed:
GH-3047: Add method and args headers into gateway

Fixes https://github.com/spring-projects/spring-integration/issues/3047

In some use-cases we can't use `headerExpressions` for messaging gateway
interface, e.g. when we use third-party class and don't use XML, but
Java DSL.

* Always add `gatewayMethod` and `gatewayArgs` headers to messages
produced by the gateway calls.
This information may be useful for downstream routing or analyzes
* Demonstrate a routing approach in the Docs with Java DSL.
Expose convenient methods to access those headers in the
`StaticMessageHeaderAccessor`
* Improve `GatewayProxyFactoryBean` to determine the return type of the
method call from the interface generic types, when the `serviceInterface`
is a `java.util.function.Function`

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/StaticMessageHeaderAccessor.java:
```diff
@@ -113,4 +114,28 @@ public static AcknowledgmentCallback getAcknowledgmentCallback(Message<?> messag
+	 */
+	@Nullable
+	public static Method getGatewayMethod(Message<?> message) {
+		return message.getHeaders().get("gatewayMethod", Method.class);
```
Constant for these two headers?

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/gateway/GatewayProxyFactoryBean.java:
```diff
@@ -484,13 +486,17 @@ protected Object doInvoke(MethodInvocation invocation, boolean runningOnCallerTh
 		Class<?> returnType = method.getReturnType();
-		boolean shouldReturnMessage = Message.class.isAssignableFrom(returnType)
-				|| hasReturnParameterizedWithMessage(method, runningOnCallerThread);
+		if (gateway.isReturnTypeMessage == null) {
```
Package protected to avoid synthetic access method; but why is this a field there anyway - it's only used here.

artembilan replied:
Correct. I just don't see reason to have an overhead parsing generics every single method call.
So, this is lazy loaded, but cached static state for the method. It is going to be definitely the same for each method invocation.

Do you mean that `private` field access is always wrapped to some generated setter call?

garyrussell replied:
Ah - makes sense, yes.

Yes, with a private field, the compiler has to generate synthetic static getters and setters; with package protected it uses simple GETFIELD and PUTFIELD operators.

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/gateway/GatewayProxyFactoryBean.java:
```diff
@@ -821,7 +828,7 @@ protected void doStop() {
 		}
 	}
 
-	private static boolean hasReturnParameterizedWithMessage(Method method, boolean runningOnCallerThread) {
```
Any reason why?

artembilan replied:
Yeah... I wanted to move `isReturnTypeMessage` logic into the `MethodInvocationGateway`, but looks like `runningOnCallerThread` is an obstacle, so I've decided to go the lazy-load approach.
Will reinstate this `static` in the next commit.

Thanks for catching this!

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/gateway/GatewayMethodInboundMessageMapper.java:
```diff
@@ -298,6 +297,10 @@ private static Expression parsePayloadExpression(Method method) {
 							: new HashMap<>();
+
+			headersToPopulate.put("gatewayMethod", holder.getMethod());
+			headersToPopulate.put("gatewayArgs", holder.getArgs());
```
We should probably add a mechanism to suppress these so the previous behavior can be retained if needed, especially for "classic" (non-DSL) gateways.

e.g. integration property `spring.integration.gateway.suppress.default.headers.interfaces=`

artembilan replied:
??? This is just about two extra header in the message. There is no any harm in logic. More over you are going to simplify many places in your config with those extra configurations for `method` header and so on.
It is already there, so config is simpler ever if that is non-DSL one.
This is what we agreed in private discussion. 
In other places we populate source-based headers unconditionally, so what is the problem here?
If you don't need then downsrteam, you can just `filter` them!

garyrussell replied:
My real concern is adding overhead to serialization with large arguments where the serialization may be so far downstream that these arguments are no longer relevant. Of course `Method` is not `Serializable` so that won't be an issue.

Another option would be to set a boolean on the GPFB in `IntegrationFlows.from(Class<?> serviceInterface, String beanName)` so we don't get these for "legacy" gateways.

artembilan replied:
OK. How about a `boolean mapInternalHeaders` option on a `MessagingGateway` and `<gateway>` with `false` by default?
It is indeed going to be `true` for `IntegrationFlows.from(Class<?> serviceInterface, String beanName)`, since this is only the place where we may not have access to annotation.

garyrussell replied:
Sounds good.

garyrussell commented on src/reference/asciidoc/gateway.adoc:
```diff
@@ -175,6 +175,33 @@ However, default headers do NOT override any `@Header` annotations in the servic
 
 The gateway now also supports a `default-payload-expression`, which is applied for all methods (unless overridden).
 
+Starting with version 5.2, the invoked gateway `Method` is always presented as a `gatewayMethod` message header, alongside with the `gatewayArgs` header for the `Object[]` with actually method invocation arguments.
```
s/presented/added/
s/alongside.../along with the `gatewayArgs` header with the value being the actual method invocation arguments in an `Object[]`./

garyrussell commented on src/reference/asciidoc/gateway.adoc:
```diff
@@ -175,6 +175,33 @@ However, default headers do NOT override any `@Header` annotations in the servic
 The gateway now also supports a `default-payload-expression`, which is applied for all methods (unless overridden).
 
+Starting with version 5.2, the invoked gateway `Method` is always presented as a `gatewayMethod` message header, alongside with the `gatewayArgs` header for the `Object[]` with actually method invocation arguments.
+This information is useful, when we can't declare messaging gateway annotations on the third party interface or XML is not appropriate configuration model.
```
on a third party ... and XML is not an appropriate

garyrussell reviewed (changes requested):
Some things to consider.

artembilan commented:
Noticed funny thing. We don't need this:
```
private StandardEvaluationContext createMethodInvocationEvaluationContext(Object[] arguments) {
	StandardEvaluationContext context = ExpressionUtils.createStandardEvaluationContext(this.beanFactory);
	context.setVariable("args", arguments);
	context.setVariable("gatewayMethod", this.method);
	return context;
}
```

The `MethodArgsHolder` can be just propagated as a `rootObject`. So, we are going to have a single `EvaluationContext`. Better for performance again!
But I guess we need to stay with "per-call" until `5.3`: even if we deprecate those SpEL variabled, we can't remove then now for backward compatibility. 😢

garyrussell commented:
>`MethodArgsHolder`

Yes, that would be a good improvement.

artembilan committed:
* Add `mapInternalHeaders` option to `@MessagingGateway`
* Add constants for `gatewayMethod` and `gatewayArgs` headers
* Propagate `MethodArgsHolder` as a `rootObject` for SpEL evaluations

artembilan committed:
* Add `@MessagingGateway.mapInternalHeaders()` option
* Make it as `false` by default for XML and `@MessagingGateway` to
avoid possible overhead
* Make it `true` for `IntegrationFlow` when we can't place
`@MessagingGateway` on the target service interface,
e.g. `java.util.function.Function`
* Deprecate `#gatewayMethod` and `#args` evaluation context variables
in favor of `MethodArgsHolder` as root object.
They wil lbe removed in the future release and a single
`EvaluationContext` will be used for all the gateway expressions

olegz commented on spring-integration-core/src/main/java/org/springframework/integration/StaticMessageHeaderAccessor.java:
```diff
@@ -113,4 +114,28 @@ public static AcknowledgmentCallback getAcknowledgmentCallback(Message<?> messag
+	@Nullable
+	public static Method getGatewayMethod(Message<?> message) {
+		return message.getHeaders().get(IntegrationMessageHeaderAccessor.GATEWAY_METHOD, Method.class);
+	}
```
I am wondering how important it is to store reference to the actual method? Reading the docs it appears that the intention is to _know the method name_. Plus having reference to a method exposes the possibility that someone may actually invoke it and consequences and so on. . .

garyrussell replied:
The problem there is you can have overloaded methods with the same name and different return types. This was a problem we fixed long ago with the #gatewayMethod SpEL variable.I had similar concerns back then but we decided the benefits outweigh the risks.

garyrussell commented:
One more thought... 

     .from(Class<?> serviceInterface,  Gatewayheaderspec headerSpec)

Where the headerSpec supplies header definitions like we do with XML and annotations?

I think that would be more consistent.

olegz commented on spring-integration-core/src/main/java/org/springframework/integration/gateway/GatewayMethodMetadata.java:
```diff
@@ -60,7 +65,9 @@ public void setPayloadExpression(String payloadExpression) {
 
 	public void setHeaderExpressions(Map<String, Expression> headerExpressions) {
-		this.headerExpressions = headerExpressions;
+		Assert.notNull(headerExpressions, "'headerExpressions' must not be null");
```
Do we really care about it being null or should we simply `clear()` and skip `putAll` if null?

olegz commented on spring-integration-core/src/main/resources/org/springframework/integration/config/spring-integration-5.2.xsd:
```diff
@@ -821,14 +821,26 @@
+					<xsd:documentation>
+						The flag to control the mapping of internal headers (e.g. 'gatewayMethod', 'gatewayArgs')
+						They are not mapped by default to avoid performance overhead in the downstream processing and/or
+						(de)serialization, when 'gatewayArgs' might be as big objects.
```
`... when 'gatewayArgs' might represent big objects.`

artembilan commented:
OK. After debating with myself a little bit I agree with Gary:

>  .from(Class<?> serviceInterface,  Gatewayheaderspec headerSpec)

So, let's take headers back, but allow end-users to configure them manually!
Even if original idea was to add a "source of data" into headers like we do in (S)FTP, HTTP, AMQP, Kafka etc., but since nobody is happy to have all the `args` in headers unconditionally, we can stay with the current behavior, but also expand it into DSL as well.

Thank you all for the feedback!
Stay tuned and I hope the solution will be there for tomorrow's release, so you will be able to review it one more time in the morning 😄

artembilan committed:
* Remove gateway internal headers population
* Introduce an
`IntegrationFlows.from(Class<?> serviceInterface, Consumer<GatewayProxySpec> endpointConfigurer)`
to allow to configure any valid gateway proxy options similar to what
we have with the `<gateway>` and `@MessagingGateway`.
This way we are very close to consistency between different approaches

artembilan commented:
So, here you are a new vision based on the `GatewayProxySpec` instead of an implicit, suspicious headers.

I'll polish Docs after your review.

Thank you!

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/config/MessagingGatewayRegistrar.java:
```diff
@@ -181,9 +184,11 @@ else if (StringUtils.hasText(asyncExecutor)) {
 	 * @return The captured values.
 	 */
-	private List<MultiValueMap<String, Object>> captureMetaAnnotationValues(AnnotationMetadata importingClassMetadata) {
+	private  static List<MultiValueMap<String, Object>> captureMetaAnnotationValues(
```
Double space

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/gateway/GatewayMethodMetadata.java:
```diff
@@ -30,37 +31,42 @@
 	}
 
-	public void setPayloadExpression(String payloadExpression) {
+	public void setPayloadExpression(Expression payloadExpression) {
```
Breaking change?

artembilan replied:
I realized that, but this is about a new `5.2` generation and there is a little chance that this class is used somewhere outside of the framework.
Plus a workaround is easy as well.
Also we can't have overloaded version with `String` because XML parser with its Java Beans introspection is going to fail.

garyrussell replied:
Ok. Agree.

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/gateway/GatewayProxyFactoryBean.java:
```diff
@@ -563,7 +572,8 @@ private Object sendOrSendAndReceive(MethodInvocation invocation, MethodInvocatio
-	private void rethrowExceptionCauseIfPossible(Throwable originalException, Method method) throws Throwable { // NOSONAR
+	private void rethrowExceptionCauseIfPossible(Throwable originalException, Method method)
+			throws Throwable { // NOSONAR
 		Class<?>[] exceptionTypes = method.getExceptionTypes();
```
Newline before

garyrussell reviewed (changes requested):
Minor issues, except possible breaking change.

garyrussell commented:
Perhaps we should drop the `default` prefix since there are no method-level overrides in this context; or, add a TODO if we think we might add method level overrides in the future if the community requests it.

artembilan commented:
> I may drag in other people for some clarifications.

So, already then we rename them back with `default` 😄

garyrussell commented:
I don't see that text you are quoting so I don't understand your comment. I am on my phone.

artembilan commented:
Sorry, I didn't notice that I hadn't copy/pasted your comment.
My point is that I agree to rename them now without `default`.

Stay tuned. Will come back soon with commit...

olegz commented on spring-integration-core/src/main/java/org/springframework/integration/gateway/GatewayProxyFactoryBean.java:
```diff
@@ -842,11 +853,24 @@ private static boolean hasReturnParameterizedWithMessage(Method method, boolean
 	}
 
+	private boolean hasReturnMessageTypeOnFunction(Method method) {
+		if (Function.class.isAssignableFrom(this.serviceInterface) && "apply".equals(method.getName())) {
```
Wondering if NOT `&& "apply".equals(method.getName())` should really trigger the exception instead of simply returning `false`? I mean that would be some gross miss-configuration, right?

artembilan replied:
Your interface may `extends Function`, but have more methods declared.
Why is that so invalid?

olegz replied:
What I am saying is that at that moment it is really not a Function. What I mean is that you may extend and have some additional methods, but in the Functional world the only method I know is `apply` and that is my entry point if I down-cast everything to Function.
It may not be a problem in your context, but i wanted to bring it up to ensure you thought of that.

artembilan replied:
OK. Thanks for your feedback, but for extra safety I would stay with this extra protection.

olegz commented on spring-integration-core/src/main/java/org/springframework/integration/gateway/GatewayProxyFactoryBean.java:
```diff
@@ -842,11 +853,24 @@ private static boolean hasReturnParameterizedWithMessage(Method method, boolean
+					ResolvableType.forClass(Function.class, this.serviceInterface)
+							.getGeneric(1)
+							.getRawClass();
+			return returnType != null && Message.class.isAssignableFrom(returnType);
```
Again, how would it be possible to have `returnType` null on the Function.apply()?

artembilan replied:
This is syntax sugar to make our IDE and Sonar happy.
this is just because `ResolvableType.getRawClass()` is `@Nullable`, so sending `null` as argument into `isAssignableFrom()` is bad practice.

artembilan committed:
* Remove `default` prefix from `GatewayProxySpec` options
* Document the change

artembilan commented:
Pushed Docs and removed `default` from the `GatewayProxySpec`.

Thank you for your review!

olegz commented:
+1 to merge!
